### DISCUSSION TITLE
Asset Key and Asset Groups impact on asset resolution

### DISCUSSION THREAD

**Discussion by ChristianKranemann** [dagster-io/dagster#15292] [Q&A]
Hey Dagster Community, 

i am having troubles understanding the relationship between asset keys and groups. 

Consider the following example:

```
from dagster import asset, Definitions, SourceAsset, AssetKey

# # works
# source_asset = SourceAsset(
#     key=AssetKey("source_asset"),
# )

# # works
# source_asset = SourceAsset(
#     key=AssetKey("source_asset"),
#     group_name="source",
# )

# # works
# source_asset = SourceAsset(
#     key=AssetKey(["source", "source_asset"]),
# )

# does not work!
source_asset = SourceAsset(
    key=AssetKey(["source", "source_asset"]),
    group_name="source",
)

@asset
def dependent_asset(source_asset):
    return source_asset + 1


defs = Definitions(assets=[source_asset, dependent_asset])
```

The 3 versions commented out work. But for the last one i get the error: 
```
dagster._core.errors.DagsterInvalidDefinitionError: Input asset '["source_asset"]' for asset '["dependent_asset"]' is not produced by any of the provided asset ops and is not one of the provided sources
```

Why is it that 

```
source_asset = SourceAsset(
    key=AssetKey("source_asset"),
    group_name="source",
)
```
and 
```
source_asset = SourceAsset(
    key=AssetKey(["source", "source_asset"]),
    group_name="source",
)
```

produce different results? What am i missing? 

Thanks in advance!

Christian

**Comment by OwenKephart** [comment]
Hi @ChristianKranemann ! There are two different concepts at play here, the **key prefix** and the **group** of an asset.

- **key prefix**: Any asset key can consist of multiple components, following a hierarchical structure. So if you have lots of assets in snowflake database, you might structure your keys as `AssetKey(["snowflake", "<database name>", "<table name>"])`. In this case, the **key prefix** would be `["snowflake", "<database name>"]`. This hierarchical structure is generally meant to correspond to an assets physical location (i.e. the storage system it resides in)
- **group**: The asset group is a non-hierarchical concept that associates assets that share a similar usecase. For example, you might place all your assets that are created/used by the Finance team in the same group, regardless of if they live in the same storage systems or not.

In order for dependencies between assets to work, the asset keys must match. When you do:

```python
def dependent_asset(source_asset):
    return source_asset + 1
```

This declares an asset with key ["dependent_asset"] which depends on an asset with key ["source_asset"]. For assets in the same **group**, we automatically do some fuzzy matching to try to match "source_asset" to an existing key, if no asset with that exact key exists. So if we find an asset in the same group as ["dependent_asset"] which has a final component of their key as ["source_asset"], then we'll match that up to the fully-qualified asset key of ["source", "source_asset"].

However, this only works when the assets are in the same group. So if you move your SourceAsset to the "source" group, they're no longer both in the same (default) asset group, and this fuzzy matching doesn't happen.

**Reply by OwenKephart** [reply]
In short, if you want SourceAsset to have both a prefix and an asset group, then you'll want to do:

```python
source_asset = SourceAsset(
    key=AssetKey(["source", "source_asset"]),
    group_name="source",
)

@asset(ins={"source_asset": AssetIn(key_prefix="source")})
def dependent_asset(source_asset):
    source_asset + 1
```

to fully qualify the input asset key

**Comment by dridk** [comment]
i was confused in the same way. an Asset Key should be a unique identifier accessible in the same manner independently of the group name or prefix.

Something like: 

```python
@asset(ins="X": AssetIn("source/source_asset")
def dependent_asset(X):
     Pass
